File: LeetCode/0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    public List<String> letterCombinations(String digits) {
        List<String> result = new ArrayList<>();
        System.out.println(digits);
        
        if (digits.length() == 0) return result;
        
        Map<Character, Character[]> dict = new HashMap<>();
        
        dict.put('2', new Character[]{'a', 'b', 'c'});
        dict.put('3', new Character[]{'d', 'e', 'f'});
        dict.put('4', new Character[]{'g', 'h', 'i'});
        dict.put('5', new Character[]{'j', 'k', 'l'});
        dict.put('6', new Character[]{'m', 'n', 'o'});
        dict.put('7', new Character[]{'p', 'q', 'r', 's'});
        dict.put('8', new Character[]{'t', 'u', 'v'});
        dict.put('9', new Character[]{'w', 'x', 'y', 'z'});
        dfs(result, dict, digits, 0, new StringBuilder());
        
        return result;
    }
    
    void dfs(List<String> result, Map<Character, Character[]> dict, String digits, int index, StringBuilder path) {
        if (path.length() == digits.length()) {
            result.add(String.valueOf(path));
            return;
        }
        
        for (char c : dict.get(digits.charAt(index)))
            dfs(result, dict, digits, index + 1, new StringBuilder(path).append(c));
    }
}

File: LeetCode/0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.kt
class Solution {
    fun letterCombinations(digits: String): List<String> {
        val results: MutableList<String> = mutableListOf()
        
        if (digits.isEmpty()) {
            return Collections.emptyList()
        }
        
        val map: Map<Char, Array<Char>> = mutableMapOf(
            '2' to arrayOf('a', 'b', 'c'),
            '3' to arrayOf('d', 'e', 'f'),
            '4' to arrayOf('g', 'h', 'i'),
            '5' to arrayOf('j', 'k', 'l'),
            '6' to arrayOf('m', 'n', 'o'),
            '7' to arrayOf('p', 'q', 'r', 's'),
            '8' to arrayOf('t', 'u', 'v'),
            '9' to arrayOf('w', 'x', 'y', 'z')
        )
        
        fun dfs(index: Int, sb: StringBuilder) {
            if (sb.length == digits.length) {
                results.add(sb.toString())
                
                return
            }
            
            for (c in map[digits[index]]!!) {
                dfs(index + 1, StringBuilder(sb).append(c));
            }
        }
        
        dfs(0, StringBuilder())
        
        return results
    }
}